### ISSUE TITLE
Bug(C3): When THB or Desinfect is set to enabled, it's not possible to disable it again.

### ISSUE THREAD

**Issue by bdecl** [wuwentao/midea_ac_lan#399]
### HA core version

2024.11.3

### Current integration version

0.6.5

### Last known working integration version

0.0.0

### Device type and model

Heat Pump Wi-Fi Controller 171H120F (0) by Midea

### Working mobile app

Smarthome

### The detail description of bug/problem

1. description the device type and name（you may have multiple devices）
Airwell RCW30 - Heat Pump Wi-Fi Controller 171H120F (0) by Midea - C3

2. description the error action or error info (device have multiple data/features)

When the TBH or Desinfect switch is enabled then it's not possible anymore to get it disabled. I observed when the TBH switch is enabled then the desinfect automatically is also enabled and the way around.

 I think that this brings us in a situation that it's impossible to disable TBH and the Desinfect disabled at the same time. TBH and Desinfect switch remains always enabled because they are activated by each other. 

The only way to get it disabled is to use the smarthome app.

### The logs

[logsC3_TBH_cannot_be_disabled.txt](https://github.com/user-attachments/files/17922779/logsC3_TBH_cannot_be_disabled.txt)

**Comment by bdecl** [comment]
In

https://github.com/georgezhao2010/midea_ac_lan/tree/master/custom_components/midea_ac_lan/midea/devices/c3/message.py

I found this on lign 63

`disinfect = 0x04 if self.disinfect or self.tbh else 0x00`

This can maybe explain the behavior. I don't know how to change this in the wuwentao/midea_ac_lan ?

**Comment by wuwentao** [comment]
@bdecl thanks for your bug report.
it's a pip lib now and current location is :
https://github.com/midea-lan/midea-local/blob/22462e68d8cc776f5c09d9b3bfea97e595701587/midealocal/devices/c3/message.py#L103

so we still need to find out the root cause.
maybe by mypy typing issue, I'm not sure now, will check it later.

**Comment by bdecl** [comment]
The issue is that tbh and disinfect switch tries to controls the same control bit. The return value sets both switches. Once you set one of the two switches to enable (1), together with the OR function, not possible to disable the switches on the time --> `disinfect = 0x04 if self.disinfect or self.tbh` else 0x00 remains always enabled.  

I adapted the code midealocal/devices/c3/device.py as folow to solve my issue

lign 103 - I removed self.disinfect or

        `disinfect = 0x04 if self.tbh else 0x00`


lign 182 - commented lign

```
#        self.disinfect = body[data_offset + 0] & 0x20 > 0
        self.tbh = body[data_offset + 0] & 0x20 > 0
```

This makes the disinfect switch obsolete but this is not an issue since it has the same function as tbh for this C3 device. The OR function must be written in a other way. 
I tested it on my local system.

**Comment by wuwentao** [comment]
@bdecl sorry for the delay response. thank you very much for your code changes and test result share.
in fact, I'm not sure with `disinfect` and `tbh` feature changes.

need to confirm below :  with this changes, it works well for all your device `disinfect` and `tbh` feature? 

in addition, @micturkey @fagallo @AndreasTUHU @ erikgieseler, seems we also have C3 device, is this bug/feature works or exist in your C3 device with current code? 
or you can manual edit the source code as @bdecl and confirm whether it can works?
I'm not sure with the changes impact to our exist device, if it works well with our device, then we can directly commit it.

**Comment by AndreasTUHU** [comment]
> in addition, @micturkey @fagallo @AndreasTUHU @erikgieseler, seems we also have C3 device, is this bug/feature works or exist in your C3 device with current code?

I can confirm this. If I turn on either one, then both turn on and after that I can't turn off neither tbh or discinfect.

**Comment by micturkey** [comment]
> @bdecl sorry for the delay response. thank you very much for your code changes and test result share. in fact, I'm not sure with `disinfect` and `tbh` feature changes.
> 
> need to confirm below : with this changes, it works well for all your device `disinfect` and `tbh` feature?
> 
> in addition, @micturkey @fagallo @AndreasTUHU @erikgieseler, seems we also have C3 device, is this bug/feature works or exist in your C3 device with current code? or you can manual edit the source code as @bdecl and confirm whether it can works? I'm not sure with the changes impact to our exist device, if it works well with our device, then we can directly commit it.
> 
> from current code:
> 
> ```
> self.disinfect = body[data_offset + 0] & 0x20 > 0
> self.tbh = body[data_offset + 0] & 0x20 > 0
> ```
> 
> I don't think disinfect is a different value with tbh, maybe they are same one. as disinfect should be a different feature,so i don't think it should works for C3 device, and maybe we can remove disinfect and make tbh is works well, then try to support disinfect in another feature request issue/PR.

Sorry, I do not have C3 device 😂

**Comment by wuwentao** [comment]
> > in addition, @micturkey @fagallo @AndreasTUHU @erikgieseler, seems we also have C3 device, is this bug/feature works or exist in your C3 device with current code?
> 
> I can confirm this. If I turn on either one, then both turn on and after that I can't turn off neither tbh or discinfect.

yes, as `disinfect` is the same as `tbh`,  but `disinfect` should be a different feature, and only `tbh` can works.
so I will try to remove `disinfect ` and only keep `tbh` in current code.
as `disinfect` will be another new feature, we should split it in two value and action.


still need to confirm with @bdecl

**Comment by AndreasTUHU** [comment]
> in addition, @micturkey @fagallo @AndreasTUHU @erikgieseler, seems we also have C3 device, is this bug/feature works or exist in your C3 device with current code?

I can confirm this. If I turn on either one, then both turn on and after that I can't turn off neither tbh or discinfect. #

**Comment by fagallo** [comment]
> in addition, @micturkey @fagallo @AndreasTUHU @erikgieseler, seems we also have C3 device, is this bug/feature works or exist in your C3 device with current code?

I haven't tank and then I can't try. This function Is disabled for me because I don't use heat pump to do domestic hot water.

**Comment by bdecl** [comment]
I tested the TBH control for a week now. I can also confirm that the control only enables the electric heater of the tank up to the target temperature. So if the DHW target is set to 55°C then the electric heater will heat up to 55°C and will automatically disable the control when reached the target. The max DHW target is 60°C. We can not call this TBH control a disinfect function and is a limited solar load shedding since the real max temp is 70°C in the control unit. I started to search more info of the interface.

So I confirm, `disinfect ` can be removed. TBH control is not a disinfect function.

**Comment by wuwentao** [comment]
@bdecl thank you very much for your confirm and test.

1. `disinfect ` will be remove and only keep `tbh` and it should be available.

2. for what you had said the HDW max temperature, it defined in this line, you can manual change it and test it.
https://github.com/midea-lan/midea-local/blob/e1a385559aec6b1526b10a595ea9a7568a69163c/midealocal/devices/c3/__init__.py#L86
once got confirm, we can change it to the right value.

**Comment by wuwentao** [comment]
found some info from old repo: https://github.com/georgezhao2010/midea_ac_lan/issues/345
@erikgieseler maybe know it.

**Comment by erikgieseler** [comment]
Unfortunately, I can't contribute anything to this topic as the TBH is now deactivated on my heat pump and is not displayed in the app. In my opinion, however, TBH and Disinfect should remain two separate functions, as TBH heats up to a maximum of DHW temperature, while Disinfect heats up to 65 - 70 degrees according to the operating instructions. Enclosed are the C3 scripts. Maybe they will help you.
[T_0000_C3_171H120F_2023062601.txt](https://github.com/user-attachments/files/18098805/T_0000_C3_171H120F_2023062601.txt)
[T0xC3_2023120100_sm.zip](https://github.com/user-attachments/files/18098808/T0xC3_2023120100_sm.zip)

**Comment by wuwentao** [comment]
@erikgieseler thank you very much for your detail info and attachment.
we can found it's two different functions from the lua script, and we just need to confirm it with our device result.
so we can confirmed that disinfect should not available now, and only tbh is works.
current disinfect is the same as tbh, or a alias of tbh.

**Comment by wuwentao** [comment]
latest v0.6.6 should fix tbh and disinfect issue.
